azapps/kafka-backup
# TC-39 fix to bug causing backup duplicates

Duplicates were created because sink task was using batch writer and that was not committing offset until batch was not committed to s3 and it had not since there was only one message in kafka.
So after batch was committed because of time it already contained 31 copies of 1 essage since put method was called ince a minute and batch was committed after half an hour.

Since we're using batch writing we can not commit offset until entire batch is safely written to s3 so until kafka knows we've already stored messages.
I've added check if record being appended to batch has offset greater than last offset in batch. This will prevent batch from containing duplicates but will result in kafka being stuck until batch writer max time elapses. Thats why i've added mechanism to force commit in case of offset reoccurrence.

[CloudWatch query with this messge](https://eu-central-1.console.aws.amazon.com/cloudwatch/home?region=eu-central-1#logsV2:logs-insights$3FqueryDetail$3D$257E$2528end$257E$25272020-09-17T21*3a59*3a59.000Z$257Estart$257E$25272020-09-16T22*3a00*3a00.000Z$257EtimeType$257E$2527ABSOLUTE$257Etz$257E$2527Local$257EeditorString$257E$2527fields*20*40timestamp*2c*20*40message*0a*7c*20sort*20*40timestamp*20asc*0a*7c*20filter*20*40message*20like*20*22lushing*20external.user.event-5*22*20or*20*40message*20like*20*22Put*20external.user.event-5*22*20or*20*40message*20like*20*22Mark*20external.user.event-5*22*20or*20*40message*20like*20*22Commit*20external.user.event*22*0a*7c*20limit*20100$257EisLiveTail$257Efalse$257EqueryId$257E$25270b1206d9-cad5-49c5-8c6c-b3a5591eb9ce$257Esource$257E$2528$257E$2527*2fecs*2fservice*2fkafka-backup$2529$2529)

**How to test**
To test spin up dreams2-connect with this kafka-backup and storage mode S3. No backup task running.
Create topic on kafka and add two messages to it. 
Start backup task.
Check if after batch max time elapses there is json on s3 with only two messages.
